crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -409,52 +415,65 @@ impl RoomEventCache {
         Ok((events, self.inner.sender.subscribe()))
     }
 
-    /// Returns the oldest back-pagination token, that is, the one closest to
-    /// the beginning of the timeline as we know it.
+    /// Starts a back-pagination for the requested number of events.
     ///
-    /// Optionally, wait at most for the given duration for a back-pagination
-    /// token to be returned by a sync.
-    pub async fn oldest_backpagination_token(
-        &self,
-        max_wait: Option<Duration>,
-    ) -> Result<Option<PaginationToken>> {
-        self.inner.oldest_backpagination_token(max_wait).await
-    }
-
-    /// Back-paginate with the given token, if provided.
+    /// This automatically takes care of waiting for a pagination token from
+    /// sync, if we haven't done that before.
     ///
-    /// If no token has been provided, it will back-paginate from the end of the
-    /// room.
+    /// # Errors
     ///
-    /// If a token has been provided, but it was unknown to the event cache
-    /// (i.e. it's not associated to any gap in the timeline stored by the
-    /// event cache), then an error result will be returned.
-    pub async fn backpaginate(
-        &self,
-        batch_size: u16,
-        token: Option<PaginationToken>,
-    ) -> Result<BackPaginationOutcome> {
-        self.inner.backpaginate(batch_size, token).await
+    /// It may return an error if the pagination token used during
+    /// back-pagination has disappeared while we started the pagination. In
+    /// that case, it's desirable to call the method again.
+    ///
+    /// TODO(bnjbvr): that's bad API, this API should restart for us!
+    pub async fn paginate_backwards(&self, batch_size: u16) -> Result<BackPaginationOutcome> {
+        self.inner.paginate_backwards(batch_size).await
+    }
+
+    /// Test-only function to get the latest pagination token, as stored in the
+    /// room events linked list.
+    #[cfg(any(test, feature = "testing"))]
+    #[doc(hidden)]
+    pub async fn get_or_wait_for_pagination_token(&self) -> Option<String> {
+        self.inner.get_or_wait_for_pagination_token().await
+    }
+
+    /// Returns a subscriber to the pagination status used for the
+    /// back-pagination integrated to the event cache.
+    pub fn pagination_status(&self) -> Subscriber<PaginatorState> {
+        self.inner.pagination.paginator.state()
     }
 }
 
+#[derive(Debug)]
+struct RoomPaginationData {
+    /// The stateful paginator instance used for the integrated pagination.
+    paginator: Paginator,
+
+    /// Have we ever waited for a previous-batch-token from the sync? We do this
+    /// at most once per room, and we reset that upon clearing the timeline
+    /// events.
```

Comment:
If you can clarify to explain why and when this can happen, that would be lovely!

Reply:
Expanded the comment a bit.